Persistence redesign

chimp1984 opened the pull request:
Based on https://github.com/bisq-network/bisq/pull/4586
Replaces https://github.com/bisq-network/bisq/pull/4527

Attempt for a more clean commit history (still a lot). As merge was getting pretty difficult I decided to re-apply all changes again and compare with old PR code and with base of #4586. Hope no bugs entered by the challenging merge process.

Update: As #4586 is now merged to master this PR got rebased on master.

chimp1984 committed:
Refactor: Move FileUtil

chimp1984 committed:
Refactor: Move CorruptedDatabaseFilesHandler

chimp1984 committed:
Refactor: Move other 2 classes to file package

chimp1984 committed:
Refactor: Rename CorruptedDatabaseFilesHandler

chimp1984 committed:
Refactor: Rename fields and methods, return early

chimp1984 committed:
Refactor: Add removeAndBackupFile method

chimp1984 committed:
Remove awaitTermination
Remove UserThread.execute on shutdown
Rename threadname

chimp1984 committed:
Add getDefaultStorageFileName method

chimp1984 committed:
Add PersistenceManager class

chimp1984 committed:
Refactor: use FileUtil.removeAndBackupFile instead of FileManager.removeAndBackupFile

chimp1984 committed:
Remove Storage and FileManager from logback

chimp1984 committed:
Add PersistenceManager.flushAllDataToDisk to shutdown to write all pending data to disk
Use new readAllPersisted method instead of setupPersistedDataHosts. Run read tasks in a thread in parallel.

Read is only done at startup.

Remove unused restart method

chimp1984 committed:
Use PersistenceManager instead of Storage for import/export

chimp1984 committed:
Remove UserThreadMappedPersistableList and use PersistableList instead

We will not need threading support anymore once we use the new persistenceManager as serialisation is done on the userThread.

chimp1984 committed:
Remove UserThreadMappedPersistableEnvelope and use PersistableEnvelope instead

We will not need threading support anymore once we use the new persistenceManager as serialisation is done on the userThread.

chimp1984 committed:
Remove ThreadedPersistableEnvelope and use PersistableEnvelope instead

We will not need threading support anymore once we use the new persistenceManager as serialisation is done on the userThread.

chimp1984 committed:
Use PersistenceManager for StoreService (not yet supported for subclasses)

chimp1984 committed:
Use PersistenceManager for MapStoreService (not yet supported for subclasses)

chimp1984 committed:
Use PersistableNetworkPayloadStore for AppendOnlyDataStoreService

chimp1984 committed:
Update PersistableNetworkPayloadStore

chimp1984 committed:
Update ProtectedDataStoreService
Use AtomicReference instead of array, add final

chimp1984 committed:
Use PersistenceManager in HistoricalDataStoreService

chimp1984 committed:
Add missing import

chimp1984 committed:
Add final at ResourceDataStoreService
Remove clone from SequenceNumberMap.

chimp1984 committed:
Apply generics and super class methods

chimp1984 committed:
Use PersistenceManager

chimp1984 committed:
Update SignedWitness classes

chimp1984 committed:
Update AccountAgeWitnessStorage classes

chimp1984 committed:
Update TempProposalStore and ProposalStore classes

chimp1984 committed:
Update TradeStatistics2Storage classes

chimp1984 committed:
Remove lombok Delegate from PersistableList
Add final
Add setAll method
Remove Iterable interface

chimp1984 committed:
Use PersistenceManager

chimp1984 committed:
Use PersistenceManager

chimp1984 committed:
Use PersistenceManager

chimp1984 committed:
Refactor: Move classes to new package

chimp1984 committed:
Use PersistenceManager

chimp1984 committed:
Use PersistableList as base class

chimp1984 committed:
Remove storage from dispute. We do persistence from outside now.
Impl. PersistablePayload
Make openingDate final and contr param

chimp1984 committed:
Use PersistenceManager. Make disputeList final

chimp1984 committed:
Remove storage and persist calls. Remove unused methods

chimp1984 committed:
Remove persist method

chimp1984 committed:
Remove storage

chimp1984 committed:
Extend PersistableList, remvoe storage

chimp1984 committed:
Use PersistenceManager

chimp1984 committed:
Use PersistenceManager

chimp1984 committed:
Use PersistenceManager

chimp1984 committed:
Use PersistenceManager

chimp1984 committed:
Use PersistenceManager

chimp1984 committed:
Use PersistenceManager. Add setAll method

chimp1984 committed:
Use PersistenceManager

chimp1984 committed:
Use PersistenceManager. We do not persist inside openOffer anymore but from outside

chimp1984 committed:
Use PersistenceManager

chimp1984 committed:
Use PersistenceManager

chimp1984 committed:
Use PersistenceManager

chimp1984 committed:
Remove unused fields, apply new params

chimp1984 committed:
Use PersistenceManager

chimp1984 committed:
Refactor: Rename persist to requestPersistence

chimp1984 committed:
Use PersistenceManager

chimp1984 committed:
Use PersistenceManager

chimp1984 committed:
Remove FileManager ans Storage. All is replaces not by PersistenceManager

chimp1984 committed:
Apply generics

chimp1984 committed:
Remove persist method

chimp1984 committed:
Add requestPersistence call
Cleanups

chimp1984 committed:
Add missing persist calls. Add generics. Remove persist method. Add final.

chimp1984 committed:
Remove unneeded param. Add requestPersistence calls. Cleanup removeTrade and addTrade methods

chimp1984 committed:
Fix remaining issues, cleanups, add persit calls, remove unused code,...

Sorry too tired to create another 20 commits...

chimp1984 committed:
Fix wrong params

chimp1984 committed:
Cleanups, add generics, add final, remove unused code/comments/todos,

chimp1984 committed:
Use decrementAndGet instead of getAndDecrement

chimp1984 committed:
Refactor: Rename getTradesAsObservableList to getObservableList

chimp1984 committed:
Revert exit code to 1 as it was before. Not sure if there have been some scripts for seeds which expect an error code at shutdown for auto restart. I remotely remember there have been something...

chimp1984 committed:
Move methods to base class

chimp1984 committed:
Rename getClosedTradables and getFailedTrades to getObservableList

chimp1984 committed:
Remove duplicate startApplication call. Remove checkCryptoSetup call and fields used only for that.

chimp1984 committed:
Execute result handler on userThread

chimp1984 committed:
Rename writeCompleted to onWriteCompleted
Add comment
Use decrementAndGet instead of getAndDecrement

Tested cases when there is an exception at write to disk, but as we call the result handler in the finally clause we get always called onWriteCompleted, so it cannot happen that we get stuck.

chimp1984 committed:
Persist all Priority.HIGH stores at shutdown to be more safe that we did not miss any state update.
Priority.HIGH stores are all those which contain private data. Others can be rebuilt from network data or are not critical like navigationPath.

chimp1984 committed:
Enforce that clients set the priority. Remove the initialize methods which did not have priority as a param.
Make initializePersistenceManager in StorageService abstract to enforce in concrete class to define priority.

Change priorities for future renaming to a different meaning. instead of priority we want to describe the category: private data, public data,.... will come in next commit

chimp1984 committed:
Rename Priority to Source. This should make it more clear what is the intention of the usage.

Rename:
LOW to NETWORK
MID to PRIVATE_LOW_PRIO
HIGH to PRIVATE

Change delay of MID/PRIVATE_LOW_PRIO from 30 min to 2 hours (we had different datastores before using it, now its only real low prio stores)

Add comment to each enum

chimp1984 committed:
Fix rename mistakes

chimp1984 committed:
Fix test class

sqrrm commented on common/src/main/java/bisq/common/persistence/PersistenceManager.java:
```diff
@@ -0,0 +1,378 @@
+                                         PersistenceManager<?> persistenceManager) {
+        persistenceManager.shutdown();
+        openInstances.decrementAndGet();
+        if (openInstances.get() == 0) {
```
To reduce confusion, how about
```suggestion
        if (openInstances.decrementAndGet() == 0) {
```

sqrrm commented on common/src/main/java/bisq/common/persistence/PersistenceManager.java:
```diff
@@ -0,0 +1,378 @@
+    // API
+    ///////////////////////////////////////////////////////////////////////////////////////////
+
+    public void initialize(T persistable, Source sourcce) {
```
Misspelled `sourcce`

chimp1984 replied:
fixed

chimp1984 commented:
Rebased again latest master merge.

sqrrm commented on common/src/main/java/bisq/common/persistence/PersistenceManager.java:
```diff
@@ -0,0 +1,378 @@
+        // We write to disk with a delay to avoid frequent write operations. Depending on the priority those delays
+        // can be rather long.
+        if (timer == null) {
+            timer = UserThread.runPeriodically(() -> {
```
Why use `runPeriodically()` instead of `runAfter()`? The effect would be the same but now it seems the write is intended to happen periodically, but two lines later that we see that it's not.

chimp1984 replied:
Yes was from earlier version, should be runAfter.

sqrrm commented on common/src/main/java/bisq/common/proto/persistable/PersistableList.java:
```diff
@@ -17,43 +17,85 @@
+    }
+
+    public void addListener(ListChangeListener<T> listener) {
+        ((ObservableList<T>) getList()).addListener(listener);
```
This does not look safe. I think `list` should be an `ObservableList` since it's required by this interface. It's first initiated to an `ArrayList` (line 39) but I assume all implementations of `PersistableList` passes an `ObservableList` as the constructor argument.

chimp1984 replied:
Yes, was moving the listeners later to the base class. But I think best here is to enforce observeable list as it is used anyway by all subclasses.

chimp1984 replied:
After looking at subclasses its not true. many dont use ObservableList. I will move back the ObservableList methods to the classes which actually use them and remove all ObservableList casts from the base class

sqrrm commented on common/src/main/java/bisq/common/proto/persistable/PersistableList.java:
```diff
@@ -17,43 +17,85 @@
+        return false;
+    }
+
+    public boolean remove(T tradable) {
```
Naming the generic `tradable` is confusing, `item` is more neutral

chimp1984 replied:
Yes, will fix

sqrrm commented on common/src/main/java/bisq/common/proto/persistable/PersistableList.java:
```diff
@@ -17,43 +17,85 @@
+        return list.size();
+    }
+
+    public boolean contains(T thing) {
```
Rather use a uniform naming for the generic, like `item`

chimp1984 replied:
Agree.

sqrrm commented on core/src/main/java/bisq/core/app/BisqExecutable.java:
```diff
@@ -148,21 +153,30 @@ protected Injector getInjector() {
 
     protected void applyInjector() {
-        setupPersistedDataHosts(injector);
+        // Subclassed might configure classes with the injector here
```
```suggestion
        // Subclasses might configure classes with the injector here
```

chimp1984 replied:
Will fix in a new commit with other changes...

sqrrm commented:
Review needed from @ManfredKarrer to be able to merge this

chimp1984 committed:
Update common/src/main/java/bisq/common/persistence/PersistenceManager.java

Co-authored-by: sqrrm <sqrrm@users.noreply.github.com>

chimp1984 committed:
Fix typo

chimp1984 committed:
Use runAfter instead of runPeriodically

chimp1984 committed:
Fix typo

chimp1984 committed:
Add PersistableListAsObservable
Remove ObservableList methods from PersistableList
Let DisputeList and TradableList extend PersistableListAsObservable
Fix param names

ManfredKarrer reviewed (dismissed):
utACK

chimp1984 committed:
Do not use persistNow for snapshots but requestPersistence

This avoids frequent write at dao sync and better performance.

sqrrm commented on common/src/main/java/bisq/common/persistence/PersistenceManager.java:
```diff
@@ -184,8 +184,8 @@ public PersistenceManager(@Named(Config.STORAGE_DIR) File dir,
+        this.initialize(persistable, persistable.getDefaultStorageFileName(), source);
     }
 
     public void initialize(T persistable, String fileName, Source sourcce) {
```
The typo continues...

chimp1984 replied:
ups...

sqrrm commented on common/src/main/java/bisq/common/proto/persistable/PersistableListAsObservable.java:
```diff
@@ -0,0 +1,50 @@
+    public PersistableListAsObservable() {
+    }
+
+    protected PersistableListAsObservable(List<T> list) {
```
Should take `ObservableList<T>`

chimp1984 replied:
No, that is called from protobug methods and there we map to a list, will be copied over to the observable in the base class. We could use a collection here as well to make it more clear that its not setting the list.

sqrrm replied:
Right you are, since the list is created by `createList()`. No need to change, this is correct.

chimp1984 replied:
See last commit...

chimp1984 committed:
Fix typo

chimp1984 committed:
Use Collection for constructor which is called from the fromProto methods.
We do not set the collection as list but we fill the list created via the createList method.

sqrrm reviewed (approved):
ACK
